File: Python/77. Combinations.py
'''
回溯：横向从i遍历到n，纵向深度为k
'''
class Solution:
    path = []
    res = []
    def backtracking(self, n, k, start):
        if len(self.path) == k:
            self.res.append(self.path[:])
            return
        for i in range(start, n-k+2+len(self.path)):
            self.path.append(i)
            self.backtracking(n, k, i+1)
            self.path.pop()

    def combine(self, n: int, k: int) -> List[List[int]]:
        self.path = []
        self.res = []
        self.backtracking(n, k, 1)
        return self.res

File: C++/77. Combinations.cpp
/*
回溯：横向从i遍历到n，纵向深度为k
*/
class Solution {
public:
    vector<int> path;
    vector<vector<int>> res;
    void backtracking(int start, int n, int k) {
        if (path.size() == k) {
            res.push_back(path);
            return;
        }
        for (int i = start; i <= n - k + 1 + path.size(); i++) {
            path.push_back(i);
            backtracking(i + 1, n, k);
            path.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        backtracking(1, n, k);
        return res;
    }
};